### PR TITLE
Improve xwayland, batch 2

### PR DESCRIPTION
**Pull request by emersion**
See https://github.com/Cloudef/wlc/blob/master/src/xwayland/xwm.c#L135

### CODE REVIEW

**Comment by SirCmpwn** [comment]
wlr_xwayland_surface_configure should probably raise an event and let rootston set the x/y.

File: include/wlr/types/wlr_wl_shell.h [hunk 1]
```diff
@@ -106,7 +106,7 @@ void wlr_wl_shell_destroy(struct wlr_wl_shell *wlr_wl_shell);
 
 void wlr_wl_shell_surface_ping(struct wlr_wl_shell_surface *surface);
 void wlr_wl_shell_surface_configure(struct wlr_wl_shell_surface *surface,
-	uint32_t edges, int32_t width, int32_t height);
+	enum wl_shell_surface_resize edges, int32_t width, int32_t height);
```
Comment (SirCmpwn) [inline include/wlr/types/wlr_wl_shell.h]: This should be a uint32_t, these are flags iirc. Enum value types cannot have more than one value.

File: rootston/xwayland.c [hunk 2]
```diff
@@ -17,7 +17,25 @@ static void handle_destroy(struct wl_listener *listener, void *data) {
...
+	struct wlr_xwayland_surface_configure_event *event = data;
+	struct wlr_xwayland_surface *xwayland_surface = event->surface;
+
+	xwayland_surface->x = event->x;
+	xwayland_surface->y = event->y;
+	xwayland_surface->width = event->width;
+	xwayland_surface->height = event->height;
+
+	roots_surface->view->x = (double) event->x;
+	roots_surface->view->y = (double) event->y;
```
Comment (SirCmpwn) [inline rootston/xwayland.c]: Drop the spaces after `(double)`

File: rootston/xwayland.c [hunk 3]
```diff
@@ -35,15 +53,20 @@ void handle_xwayland_surface(struct wl_listener *listener, void *data) {
...
+	wl_list_init(&roots_surface->request_configure.link);
+	roots_surface->request_configure.notify = handle_configure;
+	wl_signal_add(&surface->events.request_configure,
+		&roots_surface->request_configure);
 
 	struct roots_view *view = calloc(1, sizeof(struct roots_view));
 	view->type = ROOTS_XWAYLAND_VIEW;
-	view->x = view->y = 200;
+	view->x = (double) surface->x;
+	view->y = (double) surface->y;
```
Comment (SirCmpwn) [inline rootston/xwayland.c]: Here as well

File: include/wlr/types/wlr_wl_shell.h (hunk 1 above)
Comment (emersion) [reply include/wlr/types/wlr_wl_shell.h]: I believe all values are enumerated: https://wayland.freedesktop.org/docs/html/apa.html#protocol-spec-wl_shell_surface-enum-resize (but this shouldn't be part of this PR, agreed)

File: rootston/xwayland.c (hunk 3 above)
Comment (SirCmpwn) [reply rootston/xwayland.c]: Can you also open a ticket with a list of x11 clients you've tested? We'll also fill it with clients we *should* test and try to deal with any weird ones.

**Comment by SirCmpwn** [comment]
Can you have rootston log the window title when it gets set? Also the class and such. Does it make sense to merge the title and class into one signal? Is there other metadata we should make available?

File: include/wlr/types/wlr_wl_shell.h (hunk 1 above)
Comment (SirCmpwn) [reply include/wlr/types/wlr_wl_shell.h]: You're right, no big deal then.

File: xwayland/xwm.c [hunk 4]
```diff
@@ -80,12 +89,124 @@ static bool xcb_call(struct wlr_xwm *xwm, const char *func, uint32_t line,
...
+	xcb_window_t *xid = xcb_get_property_value(reply);
+	if (xid != NULL) {
+		surface->parent = lookup_surface_any(xwm, *xid);
+	} else {
+		surface->parent = NULL;
+	}
+
+	wlr_log(L_DEBUG, "XCB_ATOM_WM_TRANSIENT_FOR: %p", xid);
+	wl_signal_emit(&surface->events.set_parent, surface);
+}
```
Comment (SirCmpwn) [inline xwayland/xwm.c]: What is the WM supposed to do with the parent? Anything?

**Comment by SirCmpwn** [comment]
LGTM overall.

File: xwayland/xwm.c (hunk 4 above)
Comment (emersion) [reply xwayland/xwm.c]: I know some compositors move the parent to foreground (but under the child) when the child is activated. Some might also add a semi-transparent overlay to the parent while it has some children. I added the signal for completeness, but that's maybe not needed (at least for now).

**Comment by emersion** [comment]
>Does it make sense to merge the title and class into one signal?

Maybe. I don't really know what compositors might do with the class. The class should only be set one time, and the title can change a lot, so I'm not sure about it. Maybe the `set_class` event is not really useful.

>Is there other metadata we should make available?

wlc also exposes:
* The state (used for fullscreen, modals)
* The PID
* The window type (tooltip, utility, DnD, dropdown, popup, combo, dialog, splash)
* A `close()` function

If you think any of these are of interest, I'll add 'em. Maybe there are things wlc doesn't implement but that we should implement.

**Comment by SirCmpwn** [comment]
Hmm, I think all of those could be of value.

**Comment by emersion** [comment]
Okay, let's add them.

**Comment by SirCmpwn** [comment]
Thanks!

**Pull request merged**